File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> st;
        int left = 0;
        int right = 0;
        int ans = 0;
        while(right < s.size())
        {
        // check whether element is found up to the end or not 
        // if not found then if condition runs anotherwise else because of element is found
            if(st.find(s[right]) == st.end() )
            {
                st.insert(s[right]);
                ans = max(ans, right-left+1);
                right++;
            }else
            {
            st.erase(s[left]);
            left++;
            }
        }
        return ans;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int maxlen = 0;
        Set<Character> charset = new HashSet<>();
        int left = 0;
        for(int right=0; right<n; right++)
        {
            if(!charset.contains(s.charAt(right)))
            {
                charset.add(s.charAt(right));
                maxlen = Math.max(maxlen, right-left+1);
            }else{
                while(charset.contains(s.charAt(right)))
                {
                    charset.remove(s.charAt(left));
                    left++;
                }
                charset.add(s.charAt(right));
            }
        }
        return maxlen;
    }
}